packages/intrepid2/src/Projection/Intrepid2_ProjectionTools.hpp

```diff
@@ -770,104 +770,170 @@ class ProjectionTools {
+        { /// workspace query
+          ordinal_type info(0);
+          value_type work[2];
+          lapack.GELS('N', 
+                      numRows, numRows, numCells,
+                      nullptr, numRows,
+                      nullptr, numRows,
+                      &work[0], lwork,
+                      &info);
+          lwork = work[0];
         }
-      }
-      else {
-        Kokkos::View<valueType**,Kokkos::LayoutLeft,HostDeviceType> pivVec("pivVec", 2*(m+n), 1);
-        Kokkos::View<valueType**,Kokkos::LayoutLeft,HostDeviceType> serialElemRhs("serialElemRhs", n+m, 1 );
-        for (ordinal_type ic = 0; ic < numCells; ic++) {
-          auto A = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(),
-              Kokkos::subview(elemMat, ic, Kokkos::ALL(), Kokkos::ALL()));
-          auto b = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(),
-              Kokkos::subview(elemRhs, ic, Kokkos::ALL()));
-          auto basisCoeffs_ = Kokkos::subview(basisCoeffs, ic, Kokkos::ALL());
-          auto serialBasisCoeffs = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(),
-              basisCoeffs_);
-
-          Kokkos::deep_copy(serialElemMat,valueType(0));  //LAPACK might overwrite the matrix
-
-          for(ordinal_type i=0; i<m+n; ++i) {
-            serialElemRhs(i,0) = b(i);
-            for(ordinal_type j=0; j<n; ++j)
-              serialElemMat(j,i) = A(j,i);
-          }
-
-          for(ordinal_type i=n; i<n+m; ++i)
-            for(ordinal_type j=0; j<n; ++j)
-              serialElemMat(i,j) = serialElemMat(j,i);
-
-          // Using GELS because the matrix can be close to singular.
-          ordinal_type info = 0;
-          lapack_.GELS('N', n+m, n+m, 1,
-              serialElemMat.data(), serialElemMat.stride_1(),
-              serialElemRhs.data(), serialElemRhs.stride_1(),
-              pivVec.data(), pivVec.extent(0),
-              &info);
-
-          if (info) {
-            std::stringstream ss;
-            ss << ">>> ERROR (Intrepid::ProjectionTools::getBasisCoeffs): "
-                << "LAPACK return with error code: "
-                << info;
-            INTREPID2_TEST_FOR_EXCEPTION( true, std::runtime_error, ss.str().c_str() );
-          }
-
-          for(ordinal_type i=0; i<n; ++i) {
-            serialBasisCoeffs(elemDof(i)) = serialElemRhs(i,0);
-          }
-          Kokkos::deep_copy(basisCoeffs_,serialBasisCoeffs);
+        
+        matrix_host_type C(do_not_init_tag("C"), numRows, numCells);
+
+        host_range_policy_type policy(0, numCells);
+        {
+          Kokkos::parallel_for
+            ("ProjectionTools::solveHost::matrixIndependentOfCell::pack",
+             policy, [=](const ordinal_type & ic) {
+              for (ordinal_type i=0;i<numRows;++i)
+                C(i,ic) = elemRhs_host(ic, i);
+            });
         }
-      }
+        {
+          /// GELS does scaling and separating QR and apply Q is not stable
```

Comment:
When the matrix is ill-conditioned, QR with pivoting or SVD would be the solution as they can filter out numeric null ranks. GELS does additional scaling in addition to the QR. So, it maintain some more digits reducing round-off errors. QR itself is better (Q is unitary and condition number only depends on R) than LU with partial pivoting (condition number is amplified by both L and U) but they are in practice equivalent. Solving ill-conditioned matrix still invert the numeric null rank and it will give unstable solutions with round-off errors. My qr implementation in KokkosKernels is different from the LAPACK version. That might give the difference but when I tested LAPACK QR, it shows the same level of errors when I use LU with partial pivoting resulting test failures.

Reply:
I see, thanks for the explanation.